components/encryption/src/manager/mod.rs

```diff
@@ -786,16 +843,168 @@ impl EncryptionKeyManager for DataKeyManager {
         fail_point!("key_manager_fails_before_delete_file", |_| IoResult::Err(
             std::io::ErrorKind::Other.into()
         ));
-        self.dicts.delete_file(fname)?;
+        // `RemoveDir` is not managed, but RocksDB may use `RenameFile` on a directory,
+        // which internally calls `LinkFile` and `DeleteFile`.
+        let path = Path::new(fname);
+        if path.is_dir() {
+            let mut iter = walkdir::WalkDir::new(path).into_iter().peekable();
+            while let Some(e) = iter.next() {
+                self.dicts.delete_file(
+                    e?.path().as_os_str().to_str().unwrap(),
+                    iter.peek().is_none(),
+                )?;
+            }
+        } else {
+            self.dicts.delete_file(fname, true)?;
+        }
         Ok(())
     }
 
     fn link_file(&self, src_fname: &str, dst_fname: &str) -> IoResult<()> {
-        self.dicts.link_file(src_fname, dst_fname)?;
+        let src_path = Path::new(src_fname);
```

Comment:
doesn't rocksdb call link_file for every files instead of directories?

Reply:
RocksDB uses RenameFile for both file and directory. We implement RenameFile as 
```rust
key_manager.link_file(src, dst);
if fs.rename_file(src, dst).is_err() {
  key_manager.delete_file(dst);
} else {
  key_manager.delete_file(src);
}
```